### PR TITLE
React capstone project

### PR DESCRIPTION
**Pull request by tajemouti**
Hi,

Here's a list of the accomplished tasks while working on this project:

1. Created react app.
2. Installed project dependencies.
3. Implemented JS and styling linters.
4. Created the application component.
5. Configured and exported Redux store.
6. Added the application routes in the App component.
7. Integrated weather and air quality metrics API.
8. Handled error messages in the application.
9. Integrated a search bar in the States and Cities pages for real-time filtering.
10. Implemented unit tests for pure functions in the Redux code.
11. Created integration tests for the application using React Testing Library.
12. Mocked API access to ensure tests do not send actual requests.
13. Deployed the application to Netlify.
14. Recorded a [loom](https://www.loom.com/share/1fc7db53062e42a68cdeff8a7530a9bd?sid=05cdc5be-0933-4b10-970e-ee80aef78270).

### CODE REVIEW

File: src/reportWebVitals.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+      getCLS(onPerfEntry);
+      getFID(onPerfEntry);
+      getFCP(onPerfEntry);
+      getLCP(onPerfEntry);
+      getTTFB(onPerfEntry);
+    });
+  }
+};
+
+export default reportWebVitals;
```
Comment (tresorsawasawa) [inline src/reportWebVitals.js]: - [ **OPTIONAL** ] To make your project looks clean and to have the small size of storage, I think it would be better if you remove/delete all the files which are not related to your project.

File: src/components/City.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+          <p>Error fetching data. Please try again.</p>
+          <button type="button" onClick={retryFetch}>Retry</button>
+        </div>
+      );
+    }
+    const { current } = city;
+    return (
+      <div className="container">
+        <nav className="navbar flex">
+          <Link to={`/${selectedState}`} className="back">Back</Link>
```
Comment (tresorsawasawa) [inline src/components/City.js]: - [ ] As required for the project to follow the [design guidelines](https://www.behance.net/gallery/31579789/Ballhead-App-(Free-PSDs)) including `Typography`(font face, size, and weight), `Layout`(composition and space between elements), I think it would be better if you improve this back button design by also add the missing `ShowLeftArrow`

File: src/components/City.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            hPa
+          </p>
+          <p>
+            Humidity:
+            {current.weather.hu}
+            %
+          </p>
+        </article>
+      </div>
+    );
```
Comment (tresorsawasawa) [inline src/components/City.js]: - [ ] As required for the project to follow the [design guidelines](https://www.behance.net/gallery/31579789/Ballhead-App-(Free-PSDs)) including `Typography`(font face, size, and weight), `Layout`(composition and space between elements), I think it would be better if you improve the alignment of your element text, rows to have some different colors(light and dark).

| Template Design | Your Design |
| ---------------------- | ------------------- |
| Screenshot from 2023-09-14 04-10-01 | Screenshot from 2023-09-14 04-11-15 |

File: src/components/States.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  return (
+    <div>
+      <nav className="navbar flex">
+        <h2>Canadian States</h2>
+        <Search onSearch={handleSearch} />
+      </nav>
+      {renderStates()}
+    </div>
+  );
+};
```
Comment (tresorsawasawa) [inline src/components/States.js]: - [ ] As required for the project to follow the [design guidelines](https://www.behance.net/gallery/31579789/Ballhead-App-(Free-PSDs)) including `Typography`(`font face`, `size`, `weight`,...), `Layout`(`composition` and `space between elements`,...), I think it would be better if you improve the alignment of your element text, cards to have some different colors(light and dark) as on the original design. I can also suggest you add the missing arrow in each card element, as this can also help users know if they are clickable.

| Template Design | Your Design |
| ---------------------- | ------------------- |
| 
Screenshot from 2023-09-14 04-09-50 | Screenshot from 2023-09-14 04-11-00 |

**Review by tresorsawasawa — CHANGES_REQUESTED** [review]
# STATUS: REQUIRED CHANGES :recycle::recycle: 

Hi @tajemouti 
Good job so far! I know you worked hard to implement this project, I really appreciate the effort you put in! But there are some issues that you still need to work on to go to the next project but you are almost there! 

almost-there

## To highlight :champagne: 
- Use dark pattern of your files :heavy_check_mark: 
- Use React Router :heavy_check_mark: 
- Connect React and Redux. :heavy_check_mark: 
- Use React components :heavy_check_mark: 

## Required changes :recycle: :recycle: 
- [ ] Check the comments under the review.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear. 
Please ping me [@tresorsawasawa](https://github.com/tresorsawasawa) when you comment so I can receive the notification or use slack with the same name._

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by skyv26 — APPROVED** [review]
## Hi @tajemouti, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit:  
Congratulations! 🎉 

## STATUS: APPROVED 🟢 



## Highlights
- Good **PR** title ✅ 
- Descriptive *PR* description ✅ 
- Implement the features ✅ 
- Nice *UI* design ✅ 
- Professional `README` template ✅ 
- Linters are green ✅ 
- Used correct `Gitflow` ✅ 
- Meaningful **Github** commits ✅ 
- Correct `.github/workflows` for linters ✅ 
- Followed *best-practices* ✅ 
- Correct **Semantics** used ✅ 
- Well-structured project ✅ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

## ⚠️  WARNING ⚠️ 
**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

## Have any doubt ❓ 
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @skyv26 in your question so I can receive the notification. You can also follow me on the below platforms

<table>
<tr>
<td>

<a href="https://github.com/skyv26">
   @skyv26
</a>

</td>
<td>

<a href="https://www.linkedin.com/in/skyv2022/">
   linkedin-skyv2022
</a>

</td>
<td>

<a href="https://twitter.com/vrma_aakash">
   @vrma_aakash
</a>

</td>
<td>

<a href="https://www.instagram.com/skybrel/">
   @skybrel
</a>

</td>
<td>

<a href="https://www.frontendmentor.io/profile/skyv26">
   @skybrel
</a>

</td>
</tr>
</table>

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
